### PR TITLE
Feature: Add AnimatedList with separators

### PR DESCRIPTION
**Pull request by Peetee06**
This PR adds `AnimatedList.separated`. A widget like an AnimatedList with animated separators. `animated_list_separated.0.dart` extends `animated_list.0.dart` to work with `AnimatedList.separated`

Related issue: https://github.com/flutter/flutter/issues/48226

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by Peetee06** [comment]
I'm unsure about how to best handle `removeItem` und `removeAllItems`. If we want to reuse `_AnimatedScrollViewState`, we need to prevent users from using those, because they don't come with a parameter for building removed separators. Throwing an exception was the best I can could come up with. 

Is there a better way to do this? Would it be preferred to re-implement `_AnimatedScrollViewState` with separators?

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Hixie) [inline examples/api/lib/widgets/animated_list/animated_list_separated.0.dart:78]: trivial nit: remove the extra space before the comma

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Hixie) [inline examples/api/lib/widgets/animated_list/animated_list_separated.0.dart:89]: to make this code clearer to new programmers, i recommend increasing _nextItem separate from inserting it into the list, so it's 100% clear whether the inserted value is the pre-increment or post-increment value.

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Hixie) [inline examples/api/lib/widgets/animated_list/animated_list_separated.0.dart:129]: nit: indentation

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:323]: nit: we prefer having the space on the end of the line before the cut

(if this is copy/pasted from some other code, bonus points for fixing that code too)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:476]: I think this would lead to confusion. Can we instead find some way to let `removeItem` be how you remove items, for consistency across the various APIs?

Comment (Hixie) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:476]: e.g. have the superclass have an `@protected` `internalRemoveItem` or some such, and then have the subclasses other than this one introduce a `removeItem` that just forwards everything to `internalRemoveItem`?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:541]: avoid else after return

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:424]: @goderbauer / @Piinks 
We probably shouldn't do it in this PR but the fact that we couldn't implement AnimatedListSeparatedState outside of the framework because _AnimatedScrollViewState is private probably means we have some magic here and should consider making _AnimatedScrollViewState public.

**Review by Piinks — COMMENTED** [review]
Hi @Peetee06 welcome! Thanks very much for contributing.

Rather than a new class, we would typically approach this with a special constructor in the existing AnimatedList class. The ListView widget is a good example of this, with its ListView.separated constructor. There is a fair bit of duplicated code between the AnimatedListSeparated here and the AnimatedList that already exists in the framework, which would mean we need to maintain both, ensuring when we fix a bug in one class, we remember to go fix it as well in the other class, and so on.

Would you be interested in implementing this feature in the existing class?

**Comment by Peetee06** [comment]
Hi @Piinks,

that was actually the first thing I tried, but I wasn't able to find a solution for the `removeItem` method. It requires a single removal builder for the `AnimatedList` but two removal builders for the AnimatedList with separators, because we need to remove the separators as well. I could implement a second method like I did already but I agree with @Hixie [here](https://github.com/flutter/flutter/pull/144899#discussion_r1536884961), that having two removal methods, where one is unused for each `AnimatedList` might lead to confusion. Do you have an idea on how to implement it as `AnimatedList.separated` while not having methods that are only used in either `AnimatedList` xor `AnimatedList.separated`? 

I like @Hixie's suggestion about replacing the superclass's removeItem with a protected member to be able to implement both a `removeItem` with a single removal builder for the default `AnimatedList` and one with two removal builders for the `AnimatedList.separated` if we would choose to stay with a separated AnimatedListSeparated class. What do you think about this refactoring? 

I would be happy to improve the code so that it has a clean, developer-friendly API.

**Comment by Piinks** [comment]
> that was actually the first thing I tried, but I wasn't able to find a solution for the removeItem method. It requires a single removal builder for the AnimatedList but two removal builders for the AnimatedList with separators, because we need to remove the separators as well. I could implement a second method like I did already but I agree with @Hixie https://github.com/flutter/flutter/pull/144899#discussion_r1536884961, that having two removal methods, where one is unused for each AnimatedList might lead to confusion. Do you have an idea on how to implement it as AnimatedList.separated while not having methods that are only used in either AnimatedList xor AnimatedList.separated?

Would the removal of the separators just be expected by the user without having to call it themselves? I feel like that's what I would expect. In that case, we could just internally remove the leftover separators when items are removed. Maybe the constructor can have the user provide the add and remove builders just for the separators then?

**Comment by Piinks** [comment]
> What do you think about this refactoring?

I think these are great avenues to explore, thanks for working on this! :)

**Comment by Piinks** [comment]
@Peetee06 is this ready for another review?

**Comment by Peetee06** [comment]
@Piinks yes! I wrote a few infos about it and two different branches with other possible solutions  in the "hackers-new" channel on discord. Here is a summary:

All in all, we now have 3 possible solutions:

Use @protected internal methods for _AnimatedScrollViewState and pass a removedSeparatorBuilder to removeItem and removeAllItems: https://github.com/flutter/flutter/pull/144899
Pass removedSeparatorBuilder to the constructor of AnimatedListSeparated. No need to make _AnimatedScrollViewState's methods protected, because they can just be overriden now: https://github.com/Peetee06/flutter/tree/feat/add_AnimatedListSeparated_with_separator_removal_builder
Same as 2, but adding the removedSeparatorBuilder to _AnimatedScrollView and adding conditional logic to _AnimatedScrollViewState's methods to handle the case of having separators or not: https://github.com/Peetee06/flutter/tree/feature/add_AnimatedList_separated

From a user's perspective, I like solution 3 the most, because the API seems the cleanest and is familiar in regards to ListView.separated.

Looking forward to hearing your opinion, or ideas on how to improve on these approaches.

**Review by Piinks — COMMENTED** [review]
Wow thank you so much for digging into this so thoroughly! Sorry it took me a while to review all of the options you presented, with sample code too, so grateful you included that. :)

I really like option 3 too, consistency across APIs is really wonderful when we can achieve it. Shall we go with 3 then?

This is a really fantastic contribution for a first PR, thank you again.

**Comment by Peetee06** [comment]
Sounds good! I will update this PR to use option 3 tomorrow. :)

**Comment by Peetee06** [comment]
@Hixie @Piinks 
The PR is now ready for review.

There is a limitation regarding removedSeparatorBuilder that is observable in the example in `animated_list_separated.0.dart`. As far as I can tell, there is no way to use the removed item for building the removed separator. The [ListModel.removeAt] method removes the item from the list, while storing it in a variable. It then uses that item in the removedItemBuilder. Because we are passing the removedSeparatorBuilder to the `AnimatedList.separated` constructor instead of `removeItem`, we don't have access to the item when the removedItemBuilder is called.

Let me know if we should consider more options, or should just go with this solution for now. 

Looking forward to your reviews! :)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:106]: Should 'zero' here be `[EdgeInsets.zero]`?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:110]: Is this another way, other than zero padding?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:261]: It looks like this is the same signature as AnimatedItemBuilder, do you think the separator builder could use the same?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:892]: Same here

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:691]: I think this is adding the separator, can you add a comment if so?
What if this is the last item in the list, does it still build a separator?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:742]: Adding comments for each of these calls to removeItem would be helpful, like 'this removes the item', 'this removes the separator', etc. :)

**Review by Piinks — COMMENTED** [review]
Thank you for the updates, and for your patience!

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Peetee06) [reply examples/api/lib/widgets/animated_list/animated_list_separated.0.dart]: Good catch! Updated.

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart]: I added comments. Do they help, or do we need to improve them further? If so, please let me know what could help make this easier to understand.

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:693]: I added a comment. Does this help? Will be happy to improve this further if needed!

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:261]: My thought was to separate them semantically with different naming for item and separator. It might be nicer to only define unique signatures? I updated the docstring for `AnimatedItemBuilder` accordingly. I also thought about a different name for it, but don't want to make this a breaking change :)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart]: I tried to update the `AnimatedItemBuilder` docstring to describe its usage for the removedSeparatorBuilder. But I'm unsure, if I was able to describe it in an easy to understand way. If we should improve it further, do you have ideas in what way?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:106]: Not sure, I copy pasted this from [AnimatedList]'s docstring :)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:110]: see answer regarding 'zero'

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Piinks) [inline examples/api/lib/widgets/animated_list/animated_list_separated.0.dart:91]: The indentation here is off, can you fix it?

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:106]: Hmm. We should probably know what these docs mean before we commit this.

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:846]:
```suggestion
///
/// This signature is also used by [AnimatedList.separated] to build its separators and
```

Comment (Piinks) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:846]: Docs here LGTM! Thanks!

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart]: They help! Thank you!

**Review by Piinks — COMMENTED** [review]
Looks like it's just a few docs nits remaining here, very nearly ready to land! :)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:106]: Depending on the use case one might want to remove only the padding for a specific side. So IMO "override with a zero [padding] property." is more fitting than "override with `EdgeInsets.zero` [padding] property."

What do you think?

File: examples/api/lib/widgets/animated_list/animated_list_separated.0.dart
Comment (Peetee06) [reply examples/api/lib/widgets/animated_list/animated_list_separated.0.dart]: fixed :)

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart:106]: SGTM! Thanks for verifying!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

Thanks for all of your hard work on this! 🎉

**Comment by Peetee06** [comment]
That's amazing! Thanks for the LGTM :) Happy to have contributed!

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:152]:
```suggestion
    required AnimatedItemBuilder super.removedSeparatorBuilder,
```

If someone puts `removedSeparatorBuilder: null` in this constructor, it might cause problems. Luckily, we can specify a type for the super parameter!

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/animated_scroll_view.dart:852]: Seeing this change is incredibly satisfying, thank you 😌👍

**Review by nate-thegrate — APPROVED** [review]
I have 1 suggestion; otherwise this LGTM!

Thanks for the fantastic contribution!

File: packages/flutter/lib/src/widgets/animated_scroll_view.dart
Comment (Peetee06) [reply packages/flutter/lib/src/widgets/animated_scroll_view.dart]: that's a great catch! Applied the change :)

**Comment by Peetee06** [comment]
A build check fails with 
```
Execution failed for task ':app:packageDebug'.
> A failure occurred while executing com.android.build.gradle.tasks.PackageAndroidArtifact$IncrementalSplitterRunnable
> java.lang.OutOfMemoryError (no error message)
````
Is that related to the broken tree? Can I do anything to fix this? OutOfMemoryError to me looks unrelated to the changes.

**Comment by nate-thegrate** [comment]
> A build check fails with
> 
> ```
> Execution failed for task ':app:packageDebug'.
> > A failure occurred while executing com.android.build.gradle.tasks.PackageAndroidArtifact$IncrementalSplitterRunnable
> > java.lang.OutOfMemoryError (no error message)
> ```
> 
> Is that related to the broken tree? Can I do anything to fix this? OutOfMemoryError to me looks unrelated to the changes.

I believe that error was a flake—we just needed to re-run the test :)

When that happens in the future, feel free to click the "Rebase branch" button, or you can wait for somebody with commit access to manually reboot the failing tests.

**Pull request merged**
